src/Arara/Process/IPC/Semaphore.php

```diff
@@ -0,0 +1,132 @@
+<?php
+
+namespace Arara\Process\IPC ;
+
+/**
+ * This class provides Interprocess communication using UNIX Semaphores
+ *
+ * Code written by Nikolay and copied into Arara/Process under the terms of
+ * the MIT license.
+ *
+ * Full credit to Nikolay for his excellent class
+ * 
+ * @license http://opensource.org/licenses/MIT
+ * @author Nikolay Bondarenko http://misterion.ru 
+ */
+class Semaphore
+{
+    /**
+     * @var resource
+     */
+    protected $mutex;
+
+    /**
+     * @var string
+     */
+    protected $file;
+
+    /**
+     * @var bool
+     */
+    protected $isAcquired;
+
+    /**
+     * Creator process pid
+     *
+     * @var int
+     */
+    protected $pid;
+
+    public function __construct($key = null)
+    {
+        $this->pid = getmypid();
+
+        $this->isAcquired = false;
+        if ($key) {
+            $semKey = $key;
+        } else {
+            $this->file = tempnam(sys_get_temp_dir(), 's');
+            $semKey = ftok($this->file, 'a');
+        }
+
+        $this->mutex = sem_get($semKey, 1); //auto_release = 1 by default
+        if (!$this->mutex) {
+            throw new \RuntimeException('Unable to create the semaphore');
+        }
+    }
+
+    public function __destruct()
+    {
+        if ($this->isAcquired()) {
+            $this->release();
+        }
+
+        if ($this->pid === getmypid()) {
+            $this->remove();
+        }
```

Comment:
Hi Henrique,

That is Nikolay's code, but my understanding is if the PHP semaphore object goes out of scope or is manually unset, then the semaphore should be released, otherwise it will deadlock other processes waiting to acquire it (there would be no other way to release the lock as the object instance is the only place holding the key for the semaphore).  

It going out of scope would in effect mean releasing the lock.  Perhaps that could be included in the documentation, but seemed sensible to me.  Did you have a different idea?

D.